libavcodec/libproresdecoder.c

```diff
@@ -0,0 +1,237 @@
+    } else { /* 12b */
+        av_log(avctx, AV_LOG_DEBUG, "Auto bitdepth precision. Use 12b decoding based on codec tag.\n");
+    }
+
+    proresdecodecctx->m_decoderContext = PROpenDecoder(1, NULL);
+    return 0;
+};
+
+int decodeAppleProResFrame(AVCodecContext *avctx, AVFrame *frame, int *got_frame, AVPacket *avpkt) {
+    ff_thread_finish_setup(avctx);
+    *got_frame = 1; // Should this be set only for +ve return value?
+    return decodePacket(avctx->priv_data, avpkt, frame, 0);
+};
+
+int decodePacket(ProResDecodeCodec* decoderctx, AVPacket* pkt, AVFrame* framePtr, bool drain) {
+    if (drain) {
+        // return AVError_create(AVErrorCode_kErrorEOF);
+        return -1; // error code to be updated
+    } else if (!pkt) {
+        // return AVError_create(AVErrorCode_kErrorEAGAIN);
+        return -2; // error code to be updated
+    }
+
+    int rowBytes = PRBytesPerRowNeededInPixelBuffer(decoderctx->m_width, decoderctx->pix_fmt, decoderctx->downscale_mode);
+    uint8_t* dataPtr = (uint8_t*)malloc(rowBytes * decoderctx->m_height);
+    if (dataPtr == NULL) {
+        // return AVError_create(AVErrorCode_kMemoryAllocationFailed, "Failed to allocate memory for dataPtr");
+        return -3; // error code to be updated
+    }
+    PRPixelBuffer destinationBuffer;
+    destinationBuffer.baseAddr = dataPtr;
+    destinationBuffer.rowBytes = rowBytes;
+    destinationBuffer.format = decoderctx->pix_fmt;
+    destinationBuffer.width = decoderctx->m_width;
+    destinationBuffer.height = decoderctx->m_height;
+    int res = PRDecodeFrame(decoderctx->m_decoderContext, pkt->data, pkt->size, &destinationBuffer, decoderctx->downscale_mode, false);
+    if (res < 0) {
+        free(dataPtr);
+        // return AVError_create(AVErrorCode_kDecoderReceiveFrameFailed, "Prores decoding failed");
+        return -4; // error code to be updated
+    }
+
+    // allocate planar YCbCr buffers (2B pixels)
+    av_frame_unref(framePtr);
+    framePtr->width  = decoderctx->m_width;
+    framePtr->height = decoderctx->m_height;
+    if (decoderctx->pix_fmt == kPRFormat_v210) {
+        framePtr->format = AV_PIX_FMT_YUV422P10LE;
+    } else if (decoderctx->pix_fmt == kPRFormat_v216) {
+        framePtr->format = AV_PIX_FMT_YUV422P16LE;
+    } else if (decoderctx->pix_fmt == kPRFormat_y416) {
+        framePtr->format = AV_PIX_FMT_YUVA444P16LE;
+    }
+    int ret = av_frame_get_buffer(framePtr, 0);
+    if (ret < 0) {
+        free(dataPtr);
+        // return AVError_create(AVErrorCode_kMemoryAllocationFailed, "Failed to allocate memory for frame");
+        return -5; // error code to be updated
+    }
+
+    memset(framePtr->data[0], 0, framePtr->linesize[0] * decoderctx->m_height); // Y
+    memset(framePtr->data[1], 0, framePtr->linesize[1] * decoderctx->m_height); // U
+    memset(framePtr->data[2], 0, framePtr->linesize[2] * decoderctx->m_height); // V 
+    
```

Comment:
The code from here on below converts the packed YUV outputs in Apple's formats (the ones listed in that include header) to more generic (or rather standardized(?)) FFmpeg supported YUV formats.
**Ques:** Should. this code be another function?